From bcb3a0943e67465e862482a906bf89d07df0daff Mon Sep 17 00:00:00 2001
From: Simao Gomes Viana <xdevs23@outlook.com>
Date: Wed, 5 Oct 2016 15:41:30 +0200
Subject: [PATCH 17/19] core: remove duplicated boot jars

This filters out duplicated boot jars automatically, preventing dex2oatd errors etc.

Change-Id: I04ba9150947d0945e98d6ef849f4e57508d076f4
Signed-off-by: Simao Gomes Viana <xdevs23@outlook.com>
Signed-off-by: blinoff82 <blinov.in@gmail.com>
---
 core/dex_preopt.mk | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/core/dex_preopt.mk b/core/dex_preopt.mk
index f76d92f9f..b1bb3347d 100644
--- a/core/dex_preopt.mk
+++ b/core/dex_preopt.mk
@@ -3,6 +3,22 @@
 #
 ####################################
 
+# Filter out duplicates
+define uniq__dx
+  $(eval seen :=)
+  $(foreach _,$1,$(if $(filter $_,${seen}),,$(eval seen += $_)))
+  ${seen}
+endef
+
+PRODUCT_BOOT_JARS := $(call uniq__dx,$(subst $(space), ,$(strip $(PRODUCT_BOOT_JARS))))
+PRODUCT_BOOT_JARS_NOPREOPT := $(call uniq__dx,$(subst $(space), ,$(strip $(PRODUCT_BOOT_JARS_NOPREOPT))))
+
+# Filter out non-preopt boot jars out of preoptable boot jars
+# this is to prevent further duplicates, as well, as strictly
+# enforcing the non-preopt rule here: non-preop boot jars are
+# not allowed to stay in normal boot jars so remove them here
+PRODUCT_BOOT_JARS := $(filter-out $(PRODUCT_BOOT_JARS_NOPREOPT),$(PRODUCT_BOOT_JARS))
+
 # list of boot classpath jars for dexpreopt
 DEXPREOPT_BOOT_JARS := $(subst $(space),:,$(PRODUCT_BOOT_JARS))
 DEXPREOPT_BOOT_JARS_MODULES := $(PRODUCT_BOOT_JARS)
-- 
2.19.1

